File: C/pilha.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_SIZE 100  // Tamanho máximo da pilha

typedef struct Stack {
    int data[MAX_SIZE];
    int top;
} Stack;

// Inicializa a pilha
void initStack(Stack* s) {
    s->top = -1;
}

// Verifica se a pilha está vazia
int isEmpty(Stack* s) {
    return s->top == -1;
}

// Verifica se a pilha está cheia
int isFull(Stack* s) {
    return s->top == MAX_SIZE - 1;
}

// Insere elemento na pilha (push)
void push(Stack* s, int value) {
    if (isFull(s)) {
        printf("Erro: pilha cheia.\n");
        return;
    }
    s->data[++(s->top)] = value;
}

// Remove elemento da pilha (pop) e retorna valor
int pop(Stack* s) {
    if (isEmpty(s)) {
        printf("Erro: pilha vazia.\n");
        return -1; // valor de erro
    }
    return s->data[(s->top)--];
}

// Retorna elemento do topo sem remover (peek)
int peek(Stack* s) {
    if (isEmpty(s)) {
        printf("Erro: pilha vazia.\n");
        return -1;
    }
    return s->data[s->top];
}

// Exibe a pilha
void display(Stack* s) {
    if (isEmpty(s)) {
        printf("Pilha vazia.\n");
        return;
    }
    printf("Pilha (topo -> base): ");
    for (int i = s->top; i >= 0; i--) {
        printf("%d ", s->data[i]);
    }
    printf("\n");
}

int main() {
    Stack stack;
    initStack(&stack);

    // Operações de teste
    push(&stack, 10);
    push(&stack, 20);
    push(&stack, 30);
    display(&stack);
    printf("Peek: %d\n", peek(&stack));
    printf("Pop: %d\n", pop(&stack));
    display(&stack);
    printf("Is empty? %s\n", isEmpty(&stack) ? "Sim" : "Não");

    // Teste de overflow
    for (int i = 0; i < MAX_SIZE; i++) {
        push(&stack, i);
    }
    // Deve exibir erro quando passar do limite
    push(&stack, 1000);

    return 0;
}

File: python/pilha.py
class Stack:
    """
    Implementação simples de pilha usando lista interna em Python.
    Métodos:
      - push(item): adiciona item ao topo
      - pop(): remove e retorna o item do topo
      - peek(): retorna o item do topo sem remover
      - is_empty(): verifica se a pilha está vazia
      - size(): retorna o número de elementos
    """
    def __init__(self):
        self._items = []

    def push(self, item):
        """Adiciona um elemento ao topo da pilha"""
        self._items.append(item)

    def pop(self):
        """Remove e retorna o elemento do topo"""
        if self.is_empty():
            raise IndexError("pop from empty stack")
        return self._items.pop()

    def peek(self):
        """Retorna o elemento do topo sem remover"""
        if self.is_empty():
            raise IndexError("peek from empty stack")
        return self._items[-1]

    def is_empty(self):
        """Verifica se a pilha está vazia"""
        return len(self._items) == 0

    def size(self):
        """Retorna o número de elementos na pilha"""
        return len(self._items)

    def __repr__(self):
        return f"Stack(top -> {list(reversed(self._items))})"

# Exemplo de uso
if __name__ == "__main__":
    stack = Stack()
    stack.push(10)
    stack.push(20)
    stack.push(30)
    print("Pilha após pushes:", stack)
    print("peek():", stack.peek())
    print("pop():", stack.pop())
    print("Pilha atual:", stack)
    print("Está vazia?", stack.is_empty())
    print("Tamanho:", stack.size())